core/state_transition.go

```diff
@@ -127,11 +128,11 @@ func NewStateTransition(evm *vm.EVM, msg Message, gp *GasPool) *StateTransition
 // the gas used (which includes gas refunds) and an error if it failed. An error always
 // indicates a core error meaning that the message would always fail for that particular
 // state and would never be accepted within a block.
-func ApplyMessage(evm *vm.EVM, msg Message, gp *GasPool) ([]byte, *big.Int, error) {
+func ApplyMessage(evm *vm.EVM, msg Message, gp *GasPool) ([]byte, *big.Int, error, error) {
 	st := NewStateTransition(evm, msg, gp)
 
 	ret, _, gasUsed, err := st.TransitionDb()
-	return ret, gasUsed, err
+	return ret, gasUsed, st.vmErr, err
```

Comment:
Just return `vmerr` from `st.TransitionDb()` instead of accessing it as a member field.